cpp/src/arrow/flight/transport/ucx/ucx_server.cc

```diff
@@ -0,0 +1,647 @@
+      }
+    }
+
+    // Clean up
+    auto status = worker->driver->Close();
+    if (!status.ok()) {
+      FLIGHT_LOG_PEER(WARNING, peer) << "Failed to close worker: " << status.ToString();
+    }
+    worker->worker.reset();
+    FLIGHT_LOG_PEER(DEBUG, peer) << "Disconnected";
+  }
+
+  void DriveConnections() {
+    while (listening_.load()) {
+      while (ucp_worker_progress(worker_conn_->get())) {
+      }
+      {
+        // Check for connect requests in queue
+        std::unique_lock<std::mutex> guard(pending_connections_mutex_);
+        while (!pending_connections_.empty()) {
+          ucp_conn_request_h request = pending_connections_.front();
+          pending_connections_.pop();
+
+          auto submitted = rpc_pool_->Submit([this, request]() { WorkerLoop(request); });
+          if (!submitted.ok()) {
+            ARROW_LOG(WARNING) << "Failed to submit task to handle client "
+                               << submitted.status().ToString();
+          }
+        }
+      }
+
+      if (!listening_.load()) break;
+      auto status = ucp_worker_wait(worker_conn_->get());
+      if (status != UCS_OK) {
+        FLIGHT_LOG(WARNING) << FromUcsStatus("ucp_worker_wait", status).ToString();
+      }
+    }
+  }
+
+  void EnqueueClient(ucp_conn_request_h connection_request) {
+    std::unique_lock<std::mutex> guard(pending_connections_mutex_);
+    pending_connections_.push(connection_request);
+    guard.unlock();
+  }
+
+  arrow::Result<std::shared_ptr<ClientWorker>> CreateWorker() {
+    auto worker = std::make_shared<ClientWorker>();
+
+    ucp_worker_params_t worker_params;
+    std::memset(&worker_params, 0, sizeof(worker_params));
+    worker_params.field_mask = UCP_WORKER_PARAM_FIELD_THREAD_MODE;
+    worker_params.thread_mode = UCS_THREAD_MODE_SINGLE;
+
+    ucp_worker_h ucp_worker;
+    auto status = ucp_worker_create(ucp_context_->get(), &worker_params, &ucp_worker);
+    RETURN_NOT_OK(FromUcsStatus("ucp_worker_create", status));
+    worker->worker.reset(new UcpWorker(ucp_context_, ucp_worker));
+
+    // Set up Active Message (AM) handler
+    ucp_am_handler_param_t handler_params;
+    std::memset(&handler_params, 0, sizeof(handler_params));
+    handler_params.field_mask = UCP_AM_HANDLER_PARAM_FIELD_ID |
+                                UCP_AM_HANDLER_PARAM_FIELD_CB |
+                                UCP_AM_HANDLER_PARAM_FIELD_ARG;
+    handler_params.id = kUcpAmHandlerId;
+    handler_params.cb = HandleIncomingActiveMessage;
+    handler_params.arg = worker.get();
+
+    status = ucp_worker_set_am_recv_handler(worker->worker->get(), &handler_params);
+    RETURN_NOT_OK(FromUcsStatus("ucp_worker_set_am_recv_handler", status));
+    return worker;
+  }
+
+  /// Callback handler. A new client has connected to the server.
+  static void HandleIncomingConnection(ucp_conn_request_h connection_request,
+                                       void* data) {
+    UcxServerImpl* server = reinterpret_cast<UcxServerImpl*>(data);
```

Comment:
We shut down the UCX objects when the server is shutdown, after which this callback shouldn't be invoked anymore. And the destructor of the server also forces shutdown (in case the server is dropped without explicit shutdown).